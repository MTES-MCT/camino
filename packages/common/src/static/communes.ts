import { z } from 'zod'

export const communeIdValidator = z.string().brand('Communes')

export type CommuneId = z.infer<typeof communeIdValidator>

export const toCommuneId = (value: string) => communeIdValidator.parse(value)

export const communeValidator = z.object({ id: communeIdValidator, nom: z.string() })
export type Commune = z.infer<typeof communeValidator>