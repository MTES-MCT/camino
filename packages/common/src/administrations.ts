import { Administration, AdministrationId, sortedAdministrations } from './static/administrations.js'
import { CommuneId } from './static/communes.js'
import { DepartementId, Departements, toDepartementId } from './static/departement.js'
import { SecteursMaritimes, getDepartementsBySecteurs } from './static/facades.js'
import { onlyUnique, isNullOrUndefined } from './typescript-tools.js'

// calcule tous les départements d'une étape
const titreEtapeAdministrationsDepartementsBuild = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): DepartementId[] => {
  if (!communes) {
    throw new Error('les communes ne sont pas chargées')
  }

  const departements = communes.map(id => toDepartementId(id))

  if (secteursMaritimes) {
    departements.push(...getDepartementsBySecteurs(secteursMaritimes))
  }

  return departements.filter(onlyUnique)
}

export const getAdministrationsLocales = (communes: CommuneId[] | undefined | null, secteursMaritimes: SecteursMaritimes[] | undefined | null): AdministrationId[] => {
  const titreDepartementsIds = titreEtapeAdministrationsDepartementsBuild(communes, secteursMaritimes)
  const titreRegionsIds = titreDepartementsIds.map(departementId => Departements[departementId].regionId).filter(onlyUnique)

  // calcule toutes les administrations qui couvrent ces départements et régions
  return sortedAdministrations.reduce((titreEtapeAdministrations: AdministrationId[], administration: Administration) => {
    // si le département ou la région de l'administration ne fait pas partie de ceux de l'étape
    const isAdministrationLocale =
      (administration.departementId && titreDepartementsIds.includes(administration.departementId)) || (administration.regionId && titreRegionsIds.includes(administration.regionId))
    // alors l'administration n'est pas rattachée à l'étape
    if (isNullOrUndefined(isAdministrationLocale) || !isAdministrationLocale) return titreEtapeAdministrations

    titreEtapeAdministrations.push(administration.id)

    return titreEtapeAdministrations
  }, [])
}
