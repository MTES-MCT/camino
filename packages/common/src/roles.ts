import { AdministrationId } from './administrations'
import { isNotNullNorUndefined } from './typescript-tools'
export const ROLES = [
  'super',
  'admin',
  'editeur',
  'lecteur',
  'entreprise',
  'bureau d’études',
  'defaut'
] as const
export type Role = typeof ROLES[number]
type UserEntreprise = { role: 'entreprise'; administrationId: undefined }
type UserBureaudEtudes = {
  role: 'bureau d’études'
  administrationId: undefined
}
type UserAdmin = { role: 'admin'; administrationId: AdministrationId }
type UserLecteur = { role: 'lecteur'; administrationId: AdministrationId }
type UserEditeur = { role: 'editeur'; administrationId: AdministrationId }
type UserSuper = { role: 'super'; administrationId: undefined }
type UserDefaut = { role: 'defaut'; administrationId: undefined }

type User =
  | { role: Role; administrationId: undefined | null | AdministrationId }
  | undefined
  | null

export const isSuper = (user: User): user is UserSuper =>
  userPermissionCheck(user, 'super')

export const isAdministration = (
  user: User
): user is UserLecteur | UserAdmin | UserEditeur =>
  isAdministrationAdmin(user) ||
  isAdministrationEditeur(user) ||
  isAdministrationLecteur(user)
export const isAdministrationAdmin = (user: User): user is UserAdmin =>
  userPermissionCheck(user, 'admin') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationEditeur = (user: User): user is UserEditeur =>
  userPermissionCheck(user, 'editeur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isAdministrationLecteur = (user: User): user is UserLecteur =>
  userPermissionCheck(user, 'lecteur') &&
  isNotNullNorUndefined(user?.administrationId)
export const isEntreprise = (user: User): user is UserEntreprise =>
  userPermissionCheck(user, 'entreprise')
export const isBureauDEtudes = (user: User): user is UserBureaudEtudes =>
  userPermissionCheck(user, 'bureau d’études')
export const isDefault = (user: User): user is UserDefaut | undefined =>
  !user || userPermissionCheck(user, 'defaut')

export const isRole = (role: Role | string | undefined | null): role is Role =>
  ROLES.includes(role)

function userPermissionCheck(user: User, role: Role) {
  return user?.role === role
}
