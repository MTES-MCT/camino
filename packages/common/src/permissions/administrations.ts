import { isAdministration, isAdministrationAdmin, isAdministrationEditeur, isSuper, User } from '../roles.js'
import { AdministrationId, Administrations, sortedAdministrations } from '../static/administrations.js'
import { Departements } from '../static/departement.js'

export const canReadActivitesTypesEmails = (user: User, administrationId: AdministrationId) => {
  if (!canReadAdministrations(user)) {
    return false
  }

  if (isSuper(user) || (isAdministration(user) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministration(user)) {
    const administration = Administrations[user.administrationId]

    const administrationIds: AdministrationId[] = []

    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      administrationIds.push(...sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id))
    }

    administrationIds.push(user.administrationId)

    return administrationIds.includes(administrationId)
  }

  return false
}

export const canReadAdministrations = (user: User) => isSuper(user) || isAdministration(user)

export const canEditEmails = (user: User, administrationId: AdministrationId): boolean => {
  if (isSuper(user) || ((isAdministrationAdmin(user) || isAdministrationEditeur(user)) && Administrations[user.administrationId].typeId === 'min')) {
    // Utilisateur super ou membre de ministère (admin ou éditeur) : tous les droits
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    // Membre d'une DREAL/DEAL vis-à-vis de la DREAL elle-même,
    // ou d'un DREAL/DEAL vis-à-vis d'une administration qui dépend d'elles
    // Admin ou éditeur : modifications
    // Admin, éditeur ou lecteur : lecture

    const administration = Administrations[user.administrationId]
    if (administration.regionId) {
      const departementIds = Object.values(Departements)
        .filter(({ regionId }) => regionId === administration.regionId)
        .map(({ id }) => id)
      // On récupère toutes les administrations départementales qui sont de la même région que l’administration régionale de l’utilisateur
      const administrationIds = sortedAdministrations.filter(({ departementId }) => departementIds.includes(departementId)).map(({ id }) => id)

      if (administration.typeId === 'dre' || administration.typeId === 'dea') {
        administrationIds.push(user.administrationId)
      }

      return administrationIds.includes(administrationId)
    }
  }

  return false
}
