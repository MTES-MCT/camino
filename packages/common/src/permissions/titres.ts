import { isTitreType, TitresTypes, TitreTypeId } from '../static/titresTypes.js'
import { DEMARCHES_TYPES_IDS, DemarcheTypeId } from '../static/demarchesTypes.js'
import { isAdministrationAdmin, isAdministrationEditeur, isBureauDEtudes, isEntreprise, isSuper, User, UserNotNull } from '../roles.js'
import { AdministrationId } from '../static/administrations.js'
import { isGestionnaire } from '../static/administrationsTitresTypes.js'
import { CommuneId } from '../static/communes.js'
import { ActivitesTypesId, sortedActivitesTypes } from '../static/activitesTypes.js'
import { activitesTypesTitresTypes } from '../static/activitesTypesTitresTypes.js'
import { Departements, toDepartementId } from '../static/departement.js'
import { Regions } from '../static/region.js'
import { activitesTypesPays } from '../static/activitesTypesPays.js'
import { canAdministrationModifyTitres } from '../static/administrationsTitresTypesTitresStatuts.js'
import { TitreStatutId } from '../static/titresStatuts.js'

export const getLinkConfig = (typeId: TitreTypeId, demarches: { demarche_type_id: DemarcheTypeId }[]): { count: 'single' | 'multiple'; typeId: TitreTypeId } | null => {
  switch (typeId) {
    case 'axm':
      return { count: 'single', typeId: 'arm' }
    case 'pxm':
      return { count: 'single', typeId: 'prm' }
  }

  const titreType = TitresTypes[typeId]

  if (titreType.typeId === 'cx' && demarches.some(({ demarche_type_id }) => demarche_type_id === DEMARCHES_TYPES_IDS.Fusion)) {
    return { count: 'multiple', typeId }
  }

  return null
}

export const canLinkTitres = (user: User, administrationIds: AdministrationId[]): boolean => {
  if (isSuper(user)) {
    return true
  }

  if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    return administrationIds.includes(user.administrationId)
  }

  return false
}

export const TITRES_TYPES_IDS_DEMAT = ['arm', 'axm']

export function assertsCanCreateTitre(user: User, titreTypeId: TitreTypeId | undefined): asserts user is UserNotNull {
  if (!isTitreType(titreTypeId) || !canCreateTitre(user, titreTypeId)) {
    throw new Error('permissions insuffisantes')
  }
}

export const canCreateTitre = (user: User, titreTypeId: TitreTypeId | null): boolean => {
  if (isSuper(user)) {
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    return isGestionnaire(user.administrationId, titreTypeId)
  } else if (isEntreprise(user) || isBureauDEtudes(user)) {
    return TITRES_TYPES_IDS_DEMAT.includes(titreTypeId)
  }

  return false
}

export const canReadTitre = (user: User, titre: { public_lecture: boolean }): boolean => {
  if (isSuper(user)) {
    return true
  }

  return titre.public_lecture
}

export const canEditTitre = (user: User, titreTypeId: TitreTypeId, titreStatutId: TitreStatutId): boolean => {
  if (isSuper(user)) {
    return true
  } else if (isAdministrationAdmin(user) || isAdministrationEditeur(user)) {
    return isGestionnaire(user.administrationId, titreTypeId) && canAdministrationModifyTitres(user.administrationId, titreTypeId, titreStatutId)
  }

  return false
}

export const canDeleteTitre = (user: User): boolean => isSuper(user)

interface TitreReduced {
  titreTypeId: TitreTypeId
  // FIXME 2023-12-20 il faudrait ajouter les façades maritimes (impact sur le daily ?)
  communes: { id: CommuneId }[]
  demarches: unknown[]
}
/**
 * Vérifie que le titre peut avoir des activités
 * @param titre - titre
 */
export const canHaveActivites = (titre: TitreReduced): boolean => {
  if (titre.demarches.length === 0) {
    return false
  }

  return sortedActivitesTypes.some(activiteType => canHaveActiviteTypeId(activiteType.id, titre))
}

/**
 * Vérifie que le titre peut recevoir un type d'activité
 * @param activiteTypeId - type d'activité
 * @param titre - titre
 */
export const canHaveActiviteTypeId = (activiteTypeId: ActivitesTypesId, titre: TitreReduced): boolean => {
  if (titre.demarches.length === 0) {
    return false
  }

  if (activitesTypesTitresTypes[activiteTypeId].some(titreTypeId => titreTypeId === titre.titreTypeId)) {
    const titrePaysIds = titre.communes
      ?.map(({ id }) => toDepartementId(id))
      .map(departementId => Departements[departementId].regionId)
      .map(regionId => Regions[regionId].paysId)

    const pays = activitesTypesPays[activiteTypeId]

    return (
      // et que le type d'activité n'est relié à aucun pays
      // ou que le type d'activite est relié à l'un des pays du titre
      !pays.length || (!!titrePaysIds?.length && pays.some(paysId => titrePaysIds.some(titrePaysId => paysId === titrePaysId)))
    )
  }

  return false
}
