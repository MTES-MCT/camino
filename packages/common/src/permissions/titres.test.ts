import { assertsCanCreateTitre, canCreateTitre, canDeleteTitre, canEditTitre, canHaveActiviteTypeId, canLinkTitres, canReadTitre, getLinkConfig } from './titres.js'
import { TitresTypesIds, TitreTypeId } from '../static/titresTypes.js'
import { ADMINISTRATION_IDS, AdministrationId } from '../static/administrations.js'
import { test, expect, describe } from 'vitest'
import { testBlankUser, TestUser } from '../tests-utils.js'
import { User } from '../roles.js'
import { entrepriseIdValidator, newEntrepriseId } from '../entreprise.js'
import { ACTIVITES_TYPES_IDS } from '../static/activitesTypes.js'
import { toCommuneId } from '../static/communes.js'
import { TitreStatutId, titresStatutsArray } from '../static/titresStatuts.js'
import { isAssociee, isGestionnaire } from '../static/administrationsTitresTypes.js'

const shouldNotBeCalled = () => Promise.reject(new Error('should not be called'))

test('getTitreFromTypeId pas de fusions', () => {
  const titreFromTypeId = TitresTypesIds.reduce<{
    [key in TitreTypeId]?: {
      count: 'single' | 'multiple'
      typeId: TitreTypeId
    }
  }>((acc, titreTypeId) => {
    const linkConfig = getLinkConfig(titreTypeId, [])
    if (linkConfig !== null) {
      acc[titreTypeId] = linkConfig
    }

    return acc
  }, {})
  expect(titreFromTypeId).toMatchSnapshot()
})
test('getTitreFromTypeId fusions', () => {
  const titreFromTypeId = TitresTypesIds.reduce<{
    [key in TitreTypeId]?: {
      count: 'single' | 'multiple'
      typeId: TitreTypeId
    }
  }>((acc, titreTypeId) => {
    const fusion = getLinkConfig(titreTypeId, [{ demarche_type_id: 'fus' }])
    if (fusion !== null) {
      acc[titreTypeId] = fusion
    }

    return acc
  }, {})
  expect(titreFromTypeId).toMatchSnapshot()
})

test.each<[TestUser, AdministrationId[], boolean]>([
  [{ role: 'super' }, [], true],
  [{ role: 'admin', administrationId: ADMINISTRATION_IDS['DREAL - AUVERGNE-RHÔNE-ALPES - SIÈGE DE LYON'] }, [ADMINISTRATION_IDS['DREAL - AUVERGNE-RHÔNE-ALPES - SIÈGE DE LYON']], true],
  [{ role: 'admin', administrationId: ADMINISTRATION_IDS['DREAL - AUVERGNE-RHÔNE-ALPES - SIÈGE DE LYON'] }, [ADMINISTRATION_IDS['DREAL - PAYS DE LA LOIRE']], false],
  [{ role: 'defaut' }, [ADMINISTRATION_IDS['DREAL - AUVERGNE-RHÔNE-ALPES - SIÈGE DE LYON']], false],
])('un utilisateur $user peut modifier les liaisons d’un titre: $can ', (user, administrationIds, can) => {
  expect(canLinkTitres({ ...user, ...testBlankUser }, administrationIds)).toBe(can)
})

describe('canCreateTitre', () => {
  test('vérifie si une entreprise peut créer un titre de type', () => {
    const result: { [key in TitreTypeId]?: true } = {}

    const user: User = { role: 'entreprise', entreprises: [], ...testBlankUser }
    for (const titreTypeid of TitresTypesIds) {
      const itCan = canCreateTitre(user, titreTypeid)
      if (itCan) {
        result[titreTypeid] = itCan
        assertsCanCreateTitre(user, titreTypeid)
      } else {
        expect(() => assertsCanCreateTitre(user, titreTypeid)).toThrowError('permissions insuffisantes')
      }
    }
    expect(result).toMatchSnapshot()
  })
  test.each<TitreTypeId>(TitresTypesIds)('vérifie si un utilisateur super peut créer un titre de type %p', titreTypeId => {
    const user: User = { role: 'super', ...testBlankUser }
    expect(canCreateTitre(user, titreTypeId)).toBe(true)
    expect(() => assertsCanCreateTitre(user, titreTypeId)).not.toThrowError()
  })

  test.each<TitreTypeId>(TitresTypesIds)('vérifie si un utilisateur administrateur lecteur ne peut pas créer de titre de type %p', titreTypeId => {
    const user: User = { role: 'lecteur', administrationId: ADMINISTRATION_IDS['DEAL - MARTINIQUE'], ...testBlankUser }
    expect(canCreateTitre(user, titreTypeId)).toBe(false)
    expect(() => assertsCanCreateTitre(user, titreTypeId)).toThrowError('permissions insuffisantes')
  })

  test('vérifie si un utilisateur administrateur admin peut créer des titres', () => {
    const result: { [key in AdministrationId]?: { [key in TitreTypeId]?: true } } = {}

    Object.values(ADMINISTRATION_IDS).forEach(administrationId => {
      const user: User = { role: 'admin', administrationId, ...testBlankUser }
      for (const titreTypeid of TitresTypesIds) {
        const itCan = canCreateTitre(user, titreTypeid)
        if (itCan) {
          ;(result[administrationId] ??= {})[titreTypeid] = itCan
          assertsCanCreateTitre(user, titreTypeid)
        } else {
          expect(() => assertsCanCreateTitre(user, titreTypeid)).toThrowError('permissions insuffisantes')
        }
      }
    })

    expect(result).toMatchSnapshot()
  })
})

describe('canEditTitre', () => {
  test.each<TitreTypeId>(TitresTypesIds)('vérifie si une entreprise ne peut pas modifier un titre de type %p', titreTypeId => {
    const user: User = { role: 'entreprise', entreprises: [], ...testBlankUser }
    titresStatutsArray.forEach(titreStatut => {
      expect(canEditTitre(user, titreTypeId, titreStatut.id)).toBe(false)
    })
  })
  test.each<TitreTypeId>(TitresTypesIds)('vérifie si un utilisateur super peut créer un titre de type %p', titreTypeId => {
    const user: User = { role: 'super', ...testBlankUser }
    titresStatutsArray.forEach(titreStatut => {
      expect(canEditTitre(user, titreTypeId, titreStatut.id)).toBe(true)
    })
  })

  test.each<TitreTypeId>(TitresTypesIds)('vérifie si un utilisateur administrateur lecteur ne peut pas modifier de titre de type %p', titreTypeId => {
    const user: User = { role: 'lecteur', administrationId: ADMINISTRATION_IDS['DEAL - MARTINIQUE'], ...testBlankUser }
    titresStatutsArray.forEach(titreStatut => {
      expect(canEditTitre(user, titreTypeId, titreStatut.id)).toBe(false)
    })
  })

  test('vérifie si un utilisateur administrateur admin peut modifier des titres', () => {
    const result: { [key in AdministrationId]?: { [key in TitreTypeId]?: { [key in TitreStatutId]?: true } } } = {}

    Object.values(ADMINISTRATION_IDS).forEach(administrationId => {
      const user: User = { role: 'admin', administrationId, ...testBlankUser }
      for (const titreTypeid of TitresTypesIds) {
        for (const titreStatutId of titresStatutsArray) {
          const itCan = canEditTitre(user, titreTypeid, titreStatutId.id)
          if (itCan) {
            ;((result[administrationId] ??= {})[titreTypeid] ??= {})[titreStatutId.id] = itCan
          }
        }
      }
    })

    expect(result).toMatchSnapshot()
  })
})

test('canDeleteTitre', () => {
  expect(canDeleteTitre({ role: 'super', ...testBlankUser })).toEqual(true)
  expect(canDeleteTitre({ role: 'admin', administrationId: 'min-mtes-dgaln-01', ...testBlankUser })).toEqual(false)
  expect(canDeleteTitre({ role: 'editeur', administrationId: 'min-mtes-dgaln-01', ...testBlankUser })).toEqual(false)
  expect(canDeleteTitre({ role: 'lecteur', administrationId: 'min-mtes-dgaln-01', ...testBlankUser })).toEqual(false)
  expect(canDeleteTitre({ role: 'entreprise', entreprises: [{ id: newEntrepriseId('entrepriseId') }], ...testBlankUser })).toEqual(false)
  expect(canDeleteTitre({ role: 'bureau d’études', entreprises: [{ id: newEntrepriseId('entrepriseId') }], ...testBlankUser })).toEqual(false)
  expect(canDeleteTitre({ role: 'defaut', ...testBlankUser })).toEqual(false)
})

describe('canHaveActivites', () => {
  describe("vérifie si un titre a des activités d'un certain type", () => {
    test("ne retourne aucun type d'activité relié à un pays sur un titre sans pays", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport trimestriel d'exploitation d'or en Guyane"], {
          titreTypeId: 'axm',
          demarches: [{}],
          communes: [],
          secteursMaritime: [],
        })
      ).toEqual(false)
    })

    test("retourne un type d'activité sur un titre AXM de Guyane", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport trimestriel d'exploitation d'or en Guyane"], {
          titreTypeId: 'axm',
          demarches: [{}],
          communes: [{ id: toCommuneId('97300') }],
          secteursMaritime: [],
        })
      ).toEqual(true)
    })

    test("ne retourne aucun type d'activité sur un titre AXM de métropole", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport trimestriel d'exploitation d'or en Guyane"], {
          titreTypeId: 'axm',
          demarches: [{}],
          communes: [{ id: toCommuneId('72000') }],
          secteursMaritime: [],
        })
      ).toEqual(false)
    })

    test("retourne un type d'activité sur un titre AXM de métropole", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS['rapport d’intensité d’exploration'], {
          titreTypeId: 'prm',
          demarches: [{}],
          communes: [{ id: toCommuneId('72000') }],
          secteursMaritime: [],
        })
      ).toEqual(true)
    })

    test("ne retourne aucun type d'activité de titre AXM Guyane sur un titre PRM de métropole", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport trimestriel d'exploitation d'or en Guyane"], {
          titreTypeId: 'prm',
          demarches: [{}],
          communes: [{ id: toCommuneId('72000') }],
          secteursMaritime: [],
        })
      ).toEqual(false)
    })

    test("retourne un type d'activité de titre  qui n'a pas de pays et qui est liée à un type de titre", () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport d'exploitation (permis et concessions W)"], {
          titreTypeId: 'pxw',
          demarches: [{}],
          communes: [],
          secteursMaritime: [],
        })
      ).toEqual(true)
    })
    test('un titre sans démarche ne peut pas avoir d’activité', () => {
      expect(
        canHaveActiviteTypeId(ACTIVITES_TYPES_IDS["rapport d'exploitation (permis et concessions W)"], {
          titreTypeId: 'pxw',
          demarches: [],
          communes: [],
          secteursMaritime: [],
        })
      ).toEqual(false)
    })
  })
})

describe('canReadTitre', () => {
  test('en tant que super je peux lire tous les titres', async () => {
    expect(await canReadTitre({ ...testBlankUser, role: 'super' }, shouldNotBeCalled, shouldNotBeCalled, shouldNotBeCalled, { public_lecture: false })).toBe(true)
  })

  test('si le titre est en lecture publique, tout le monde y a accès', async () => {
    expect(await canReadTitre(null, shouldNotBeCalled, shouldNotBeCalled, shouldNotBeCalled, { public_lecture: true })).toBe(true)
  })

  describe("pour les utilisateurs administrations, on peut lire un titre si l'utilisateur fait partie d'une administration gestionnaire, associé ou locale", () => {
    test('gestionnaire', async () => {
      const adminId = 'dea-guyane-01'
      const titreTypeId = 'axm'
      expect(isGestionnaire(adminId, titreTypeId)).toBe(true)
      expect(isAssociee(adminId, titreTypeId)).toBe(false)
      expect(
        await canReadTitre(
          { ...testBlankUser, role: 'admin', administrationId: adminId },
          () => Promise.resolve(titreTypeId),
          () => Promise.resolve([]),
          shouldNotBeCalled,
          { public_lecture: false }
        )
      ).toBe(true)
    })
    test('associée', async () => {
      const adminId = 'dea-guyane-01'
      const titreTypeId = 'arm'
      expect(isGestionnaire(adminId, titreTypeId)).toBe(false)
      expect(isAssociee(adminId, titreTypeId)).toBe(true)
      expect(
        await canReadTitre(
          { ...testBlankUser, role: 'admin', administrationId: adminId },
          () => Promise.resolve(titreTypeId),
          () => Promise.resolve([]),
          shouldNotBeCalled,
          { public_lecture: false }
        )
      ).toBe(true)
    })
    test('locale', async () => {
      const adminId = 'dre-aura-01'
      const titreTypeId = 'cxm'
      expect(isGestionnaire(adminId, titreTypeId)).toBe(false)
      expect(isAssociee(adminId, titreTypeId)).toBe(false)
      expect(
        await canReadTitre(
          { ...testBlankUser, role: 'admin', administrationId: adminId },
          () => Promise.resolve(titreTypeId),
          () => Promise.resolve([adminId]),
          shouldNotBeCalled,
          { public_lecture: false }
        )
      ).toBe(true)
    })

    test('ni gestionnaire, ni associée, ni locale', async () => {
      const adminId = 'dre-aura-01'
      const titreTypeId = 'cxm'
      expect(isGestionnaire(adminId, titreTypeId)).toBe(false)
      expect(isAssociee(adminId, titreTypeId)).toBe(false)
      expect(
        await canReadTitre(
          { ...testBlankUser, role: 'admin', administrationId: adminId },
          () => Promise.resolve(titreTypeId),
          () => Promise.resolve([]),
          shouldNotBeCalled,
          { public_lecture: false }
        )
      ).toBe(false)
    })
  })

  describe("pour les utilisateurs entreprises, on peut lire un titre si l'utilisateur fait partie d'une entreprise titulaire ou amodiataire", () => {
    test('entreprise titulaire', async () => {
      const entrepriseId = entrepriseIdValidator.parse('entrepriseId')
      expect(
        await canReadTitre({ ...testBlankUser, role: 'entreprise', entreprises: [{ id: entrepriseId }] }, shouldNotBeCalled, shouldNotBeCalled, () => Promise.resolve([entrepriseId]), {
          public_lecture: false,
        })
      ).toBe(true)
    })

    test('entreprise non titulaire', async () => {
      const entrepriseIdTitulaire = entrepriseIdValidator.parse('entrepriseIdTitulaire')
      const entrepriseIdNonTitulaire = entrepriseIdValidator.parse('entrepriseIdNonTitulaire')
      expect(
        await canReadTitre(
          { ...testBlankUser, role: 'entreprise', entreprises: [{ id: entrepriseIdTitulaire }] },
          shouldNotBeCalled,
          shouldNotBeCalled,
          () => Promise.resolve([entrepriseIdNonTitulaire]),
          { public_lecture: false }
        )
      ).toBe(false)
    })
  })
})
