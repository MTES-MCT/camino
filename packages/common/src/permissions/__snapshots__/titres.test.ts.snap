// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTitreFromTypeId fusions 1`] = `
Object {
  "apc": null,
  "aph": null,
  "apm": null,
  "apw": null,
  "arc": null,
  "arg": null,
  "arm": null,
  "axm": Object {
    "count": "single",
    "typeId": "arm",
  },
  "cxf": Object {
    "count": "multiple",
    "typeId": "cxf",
  },
  "cxg": Object {
    "count": "multiple",
    "typeId": "cxg",
  },
  "cxh": Object {
    "count": "multiple",
    "typeId": "cxh",
  },
  "cxi": Object {
    "count": "multiple",
    "typeId": "cxi",
  },
  "cxm": Object {
    "count": "multiple",
    "typeId": "cxm",
  },
  "cxr": Object {
    "count": "multiple",
    "typeId": "cxr",
  },
  "cxs": Object {
    "count": "multiple",
    "typeId": "cxs",
  },
  "cxw": Object {
    "count": "multiple",
    "typeId": "cxw",
  },
  "inh": null,
  "ini": null,
  "inm": null,
  "inr": null,
  "pcc": null,
  "prf": null,
  "prg": null,
  "prh": null,
  "pri": null,
  "prm": null,
  "prr": null,
  "prs": null,
  "prw": null,
  "pxf": null,
  "pxg": null,
  "pxh": null,
  "pxi": null,
  "pxm": Object {
    "count": "single",
    "typeId": "prm",
  },
  "pxr": null,
  "pxw": null,
}
`;

exports[`getTitreFromTypeId pas de fusions 1`] = `
Object {
  "apc": null,
  "aph": null,
  "apm": null,
  "apw": null,
  "arc": null,
  "arg": null,
  "arm": null,
  "axm": Object {
    "count": "single",
    "typeId": "arm",
  },
  "cxf": null,
  "cxg": null,
  "cxh": null,
  "cxi": null,
  "cxm": null,
  "cxr": null,
  "cxs": null,
  "cxw": null,
  "inh": null,
  "ini": null,
  "inm": null,
  "inr": null,
  "pcc": null,
  "prf": null,
  "prg": null,
  "prh": null,
  "pri": null,
  "prm": null,
  "prr": null,
  "prs": null,
  "prw": null,
  "pxf": null,
  "pxg": null,
  "pxh": null,
  "pxi": null,
  "pxm": Object {
    "count": "single",
    "typeId": "prm",
  },
  "pxr": null,
  "pxw": null,
}
`;

exports[`vérifie si une entreprise peut créer un titre de type "apc" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "aph" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "apm" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "apw" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "arc" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "arg" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "arm" 1`] = `true`;

exports[`vérifie si une entreprise peut créer un titre de type "axm" 1`] = `true`;

exports[`vérifie si une entreprise peut créer un titre de type "cxf" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxg" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxh" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxi" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxm" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxr" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxs" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "cxw" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "inh" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "ini" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "inm" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "inr" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pcc" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prf" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prg" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prh" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pri" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prm" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prr" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prs" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "prw" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxf" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxg" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxh" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxi" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxm" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxr" 1`] = `false`;

exports[`vérifie si une entreprise peut créer un titre de type "pxw" 1`] = `false`;
